bson/bsoncodec/default_value_decoders.go

```diff
@@ -62,15 +63,88 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
 		RegisterDefaultDecoder(reflect.Struct, &StructCodec{cache: make(map[reflect.Type]*structDescription), parser: DefaultStructTagParser})
 }
 
+// getValue check if value is a pointer and return final value to
+func getValue(val reflect.Value) reflect.Value {
+	if val.Kind() == reflect.Ptr {
+		if val.IsNil() {
+			val.Set(reflect.New(val.Type().Elem()))
+		}
+		return val.Elem()
+	}
+	return val
+}
+
+// getValueKind get value's kind or it's element's kind
+func getValueKind(val reflect.Value) reflect.Kind {
+	if val.Kind() == reflect.Ptr {
+		return val.Type().Elem().Kind()
+	}
+	return val.Kind()
+}
+
+// checkTypeError check ValueReader is one of required bsontypes, an error is returned otherwise
+func checkTypeError(vr bsonrw.ValueReader, requiredType ...bsontype.Type) error {
+	currentType := vr.Type()
+	for _, t := range requiredType {
+		if currentType == t {
+			return nil
+		}
+	}
+	typeString := []string{}
+	for _, t := range requiredType {
+		var payload string
+		switch t {
+		case bsontype.Decimal128:
+			payload = "decimal.Decimal128"
+		default:
+			payload = t.String()
+		}
+		typeString = append(typeString, payload)
+	}
+	return fmt.Errorf("cannot decode %v into a %s", vr.Type(), strings.Join(typeString, ", "))
+}
+
 // BooleanDecodeValue is the ValueDecoderFunc for bool types.
 func (dvd DefaultValueDecoders) BooleanDecodeValue(dctx DecodeContext, vr bsonrw.ValueReader, i interface{}) error {
-	if vr.Type() != bsontype.Boolean {
-		return fmt.Errorf("cannot decode %v into a boolean", vr.Type())
+	var (
+		err    error
+		b      bool
+		isNull bool
+	)
+	readValue := func() {
```

Comment:
I think we're adding a little too much indirection here. Let's instead just split off the fast path decoders into separate methods, and call them from this function if the type matches. It's fine to duplicate this specific logic between the methods.

Reply:
I am not sure I understood correctly, so I only update the Boolean case. If you think it's ok, I can work on the rest functions

Reply:
I added a patch with an updated [`BooleanDecodeValue`](https://github.com/mongodb/mongo-go-driver/commit/dc6a94e7fc31b407ad13fbd3d2e11a6ff069c76d#diff-d7870a7e54c1393f5080c1fbf611666dR114), can you update your patch to follow this design? If you want to do some things to make the code more compact, feel free.